### PR TITLE
[TASK] Refactor for usage outside TYPO3 instance

### PR DESCRIPTION
**Pull request by calien666**
This commit adds the possibility of adding storages outside public with having speaking file names in URL. To do so, the getPublicUrl() event is used to manipulate the public URL of a non-public file storage. With this, a real path is generated and not the eId for dumpFile from TYPO3 core is used. This allows a fine granulated trigger to the found storages, and checks access real with a 403 response if no logged-in user is set.

### CODE REVIEW

**Review by Riiiad — APPROVED** [review]
Code look good. 👍

File: Classes/Service/AccessService.php [hunk 1]
```diff
@@ -31,35 +34,36 @@ public static function getProtectedFileStorages(): array
...
+        // otherwise look into file storage
+        $datasetGroupArray = $feUser->getGroupIds();
+        $folderAccess = self::findDirectoryAccess($storage, $identifier);
+        $storageRecord = $storage->getStorageRecord();
+        $groupArray = $folderAccess ? $folderAccess->getFeGroup() : GeneralUtility::intExplode(',', $storageRecord['fe_group']);
 
-        // check if user has group or enabled for every login
+        // check if the user has a related group or enabled for every login
         if (
             in_array(-2, $groupArray) ||
```
Comment (sbuerk) [inline Classes/Service/AccessService.php:68]: Please, use the third argument of in_array() here to ensure struct type check,

```php
in_array(-2, $groupArray, true)
```

File: Classes/Service/AccessService.php [hunk 2]
```diff
@@ -31,35 +34,36 @@ public static function getProtectedFileStorages(): array
...
             count(array_diff($groupArray, $feUser->getGroupIds())) < count($groupArray) ||
```
Comment (sbuerk) [inline Classes/Service/AccessService.php:69]: micro-optimization:

Do the count for $groupArray into a local variable once before the if and avoid a double count of the same array. It's impossible that it changes between the two condition checks.

File: ext_tables.sql [hunk 3]
```diff
@@ -9,7 +9,7 @@ CREATE TABLE tx_securefilemount_folder
     folder      varchar(255) default ''  not null,
     folder_hash varchar(40)  DEFAULT ''  NOT NULL,
 
-    fe_group   varchar(255) default '0' not null,
+    fe_group   varchar(255) default null,
```
Comment (sbuerk) [inline ext_tables.sql:12]: Maybe rething if this column should not be named `fe_groups` in a dedicated follow up change. Not as blocker for this pull-request.

Not sure why the default / nullable change here, but buying it - assuming there is a valid reason for it.

**Review by sbuerk — CHANGES_REQUESTED** [review]

File: Classes/Service/AccessService.php (hunk 1 above)
Comment (calien666) [reply Classes/Service/AccessService.php]: Done

**Review by sbuerk — APPROVED** [review]
PR looks good for me now.

**Pull request merged**